feat: Add the support for SWIFT Business Identifier Code (BIC) format

codisart opened the pull request:
<!--
Fill in the relevant information below to help triage your issue.

Pick the target branch based on the following criteria:
  * Documentation improvement: master branch
  * Bugfix: master branch
  * QA improvement (additional tests, CS fixes, etc.) that does not change code
    behavior: master branch
  * New feature, or refactor of existing code: develop branch

You MUST provide a signoff in your commits for us to be able to accept your
patch; you can do this by providing either the --signoff or -s flag when using
"git commit". Please see the project contributing guide and
https://developercertificate.org for details.
-->

|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | no
| New Feature   | yes

### Description

Implementation proposal for #20

codisart committed:
feat: Add the support for SWIFT Business Identifier Code (BIC) format

Signed-off-by: codisart <louis.celier@gmail.com>

Xerkus commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,32 @@
+# Iban Validator
+
+`Laminas\Validator\BusinessIdentifierCode` validates if a given value could be a BIC code. BIC is
+the abbreviation for "Business Identifier Code".
```
Add ISO number defining the BIC. 
Also, I think, historical reference quote would be appropriate here
```suggestion
`Laminas\Validator\BusinessIdentifierCode` validates if a given value could be
a BIC code as defined by ISO 9362:2014. BIC is the abbreviation for "Business
Identifier Code".

> The original scope of ISO 9362 was limited to the identification of banks;
> consequently, the acronym “BIC” was used as an abbreviation for “bank
> identifier code”. The scope has since been extended, in the first place to
> cover all types of financial institutions, and in the 2009 edition, to cover
> non-financial institutions as well.
>
> --- [ISO 9362:2014 Banking — Banking telecommunication messages — Business
> identifier code (BIC)](https://www.iso.org/standard/60390.htm)
```

codisart replied:
I applied your suggestion

michalbundyra commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,41 @@
+`Laminas\Validator\BusinessIdentifierCode` validates if a given value could be
+a BIC code as defined by ISO 9362:2014. BIC is the abbreviation for "Business
+Identifier Code".
+> The original scope of ISO 9362 was limited to the identification of banks;
```
I think we need one empty line before blockquote

codisart replied:
added

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+<?php
+/**
```
we need one empty line after php open tag and before file-level doc-block (as PSR-12)

codisart replied:
added

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+{
+    const INVALID = 'valueNotBic';
+    const NOT_STRING = 'valueNotString';
+    const NOT_VALID_COUNTRY = 'valueNotCountry';
```
as we are on PHP 7.1 already we should define visibility for constants

codisart replied:
added

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+    const NOT_VALID_COUNTRY = 'valueNotCountry';
+
+    /**
+     * @var array
```
we can be more specific here, either: `string[]` or `array<string, string>`

codisart replied:
added

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * @inheritDoc
+     */
+    public function isValid($value)
```
as we are on PHP 7.1 we should add return type `: bool` here

codisart replied:
added

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+        self::NOT_VALID_COUNTRY => 'Invalid country code',
+    ];
+
+    private const REGEX_BIC = '/^[A-Za-z]{4}([A-Za-z]{2})[0-9A-Za-z]{2}([0-9A-Za-z]{3})?$/';
```
as we accept lowercase and uppercase I think it's better to use `i` modifier, and remove `A-Z` from the regexp.

also: for better readability it would be nice to name the capture group (country), so latter can be used instead of numeric index.

codisart replied:
I added the `i` modifier and the named country

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        if (empty($value)
+            || !preg_match(self::REGEX_BIC, $value, $matches)
```
minor: it looks like it's not defined in CS rules for that library but we tend to use single space after `!` operator (as you've done in line 57), the same below please - line 69.

codisart replied:
added space

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+
+    private function isSwiftValidCountry(string $countryCode) : bool
+    {
+        return in_array($countryCode, self::ISO_COUNTRIES)
```
It looks like we are missing test cases with lowecase country code (or in general lowercase BIC).
We allow these in regular expression but then we compare them with uppercase country codes.

Also: please add `true` as the 3rd parameter so we have strict type comparison.

Also: do we need separate constant for Kosovo? I see very minor advantage of it, and I am not super convinced to that solution, but I'd like to hear what you think. Thanks.

codisart replied:
My tought was to clearly separate the official iso list of the exception. I would even find natural to have an iso country validator.

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,82 @@
+        return in_array($countryCode, self::ISO_COUNTRIES)
+            || $countryCode === self::KOSOVO_EXCEPTION;
+    }
+}
```
Please add one empty line at the end of the file - it should be reported by CS checks (just noted that somehow CS check is not running on travis for that repo - will fix it shortly) - on local you can run `composer cs-check` to run checks, and `composer cs-fix` to apply some auto-fixes. For more details please see our [CONTRIBUTING guide](https://github.com/laminas/.github/blob/master/CONTRIBUTING.md).

codisart replied:
ran both

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+<?php
+/**
```
as noted before: please add one empty line after php open tag

codisart replied:
added

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+use Laminas\Validator\BusinessIdentifierCode;
+
+/**
+ * \Laminas\BusinessIdentifierCode
```
This is wrong, but in general I think we can remove this whole doc-block. No additional value provided here.

codisart replied:
Yes it was a bad copy paste

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+            'BANQUE ATLANTIQUE COTE D\'IVOIRE, ABIDJAN, Cote d\'Ivoire' => ['ATCICIAB'],
+            'BANQUE NATIONALE DU CANADA, MONTREAL, Canada' => ['BNDCCAMMINT'],
+            'BDO UNIBANK, INC., MANILA, Philippines' => ['BNORPHMM'],
+            'FEDERATION DES CAISSES DESJARDINS DU QUEBEC, LEVIS, Canada ' => ['CCDQCAMM'],
```
redundant spaces at the and of the name - here and in couple more cases below.

codisart replied:
removed

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+            'NEDBANK LIMITED, JOHANNESBURG, South Africa (primary office)' => ['NEDSZAJJXXX'],
+            'LA BANQUE POSTALE, MONTPELLIER, France' => ['PSSTFRPPMON'],
+            'LA BANQUE POSTALE, NANTES, France' => ['PSSTFRPPNTE'],
+            'UNICREDIT S.P.A., MILANO, Italy ' => ['UNCRITMM'],
```
as noted before we are missing any lowercase BIC

codisart replied:
I added some lowercase tests

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * @dataProvider successProvider
+     */
+    public function testValidateSuccess($code)
```
we can have type hint `string $code`

codisart replied:
added

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+        self::assertTrue($validator->isValid($code));
+    }
+
+    public function NotAStringProvider()
```
method name should start form lowercase letter, the same for other providers below

codisart replied:
changed

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+        $validator = new BusinessIdentifierCode();
+        self::assertFalse($validator->isValid($code));
+        self::assertCount(1, $validator->getMessages());
+        self::assertSame('Invalid type given; string expected', $validator->getMessages()['valueNotString']);
```
Should we use constant here instead of string `valueNotString` ?
The same in tests below.

codisart replied:
I have the same discussion with frosch on the update phpunit 9 PR. I think that to use the class constant in this test would add a coupling and would like `Inappropriate Intimacy` smell.

michalbundyra replied:
@froschdesign can you please have a look on that? I think it would be ok to use the constant here, but then we would need the constant to be public. What are your thoughts? Thanks!

froschdesign replied:
All constants must be public because they are used to overwrite the validation messages.

See in the documentation:

> ```php
> use Laminas\Validator\GreaterThan;
> 
> $validator = new GreaterThan();
> $validator->setMessage('Please enter a lower value', GreaterThan::NOT_GREATER);
> ```

https://docs.laminas.dev/laminas-validator/messages/#validation-messages

Therefore we can also use the constants in unit tests.

michalbundyra commented on test/BusinessIdentifierCodeTest.php:
```diff
@@ -0,0 +1,111 @@
+        self::assertCount(1, $validator->getMessages());
+        self::assertSame('Invalid country code', $validator->getMessages()['valueNotCountry']);
+    }
+}
```
As noted before: we need one empty line at the end of the file.

codisart replied:
added

michalbundyra reviewed (changes requested):
@codisart Thanks a lot for your contribution. I've added some comments.
We need also rebase and target develop as it is new feature and can be released in next minor. Thanks 👍

froschdesign commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,84 @@
+        'SG', 'SH', 'SI', 'SJ', 'SK', 'SL', 'SM', 'SN', 'SO', 'SR', 'SS', 'ST', 'SV', 'SX', 'SY', 'SZ', 'TC', 'TD',
+        'TF', 'TG', 'TH', 'TJ', 'TK', 'TL', 'TM', 'TN', 'TO', 'TR', 'TT', 'TV', 'TW', 'TZ', 'UA', 'UG', 'UM', 'US',
+        'UY', 'UZ', 'VA', 'VC', 'VE', 'VG', 'VI', 'VN', 'VU', 'WF', 'WS', 'YE', 'YT', 'ZA', 'ZM', 'ZW',
+    ];
```
I think a link to reference (list) would help here.

codisart replied:
added

froschdesign commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,84 @@
+        'UY', 'UZ', 'VA', 'VC', 'VE', 'VG', 'VI', 'VN', 'VU', 'WF', 'WS', 'YE', 'YT', 'ZA', 'ZM', 'ZW',
+    ];
+
+    private const KOSOVO_EXCEPTION = 'XK';
```
The same here, a link to reference would help.

codisart replied:
added

froschdesign replied:
I'm sorry, but I forgot to point out that we use [`@see` for links](https://docs.phpdoc.org/latest/references/phpdoc/tags/see.html).

froschdesign commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,84 @@
+        self::NOT_VALID_COUNTRY => 'Invalid country code',
+    ];
+
+    private const REGEX_BIC = '/^[a-z]{4}(?<country>[a-z]{2})[0-9a-z]{2}([0-9a-z]{3})?$/i';
```
The same here, if there are an official pattern then a link should be added.

codisart replied:
The pattern is defined in the wikipedia page. Would it be sufficient ?

froschdesign replied:
If the pattern is adopted, we should add the reference, and if the official specification is behind a paywall, Wikimedia is good.

boesing replied:
I've found a regex provided by the "Bundesbank" in this PDF:

https://www.bundesbank.de/resource/blob/749660/d2c6e00664251b4d83483c229e084e44/mL/technische-spezifikationen-scc-anhang-112018-data.pdf

Page 39. Maybe we might use that regex instead or we have a better "official" resource.

froschdesign commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,42 @@
+# Iban Validator
```
Wrong name in headline!

codisart replied:
changed

froschdesign commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,42 @@
+>
+> --- [ISO 9362:2014 Banking — Banking telecommunication messages — Business
+> identifier code (BIC)](https://www.iso.org/standard/60390.htm)
+
```
```suggestion
`Laminas\Validator\BusinessIdentifierCode` validates if a given value **could be a ["Business Identifier Code" (BIC)](https://www.swift.com/standards/data-standards/bic)** as defined by [ISO 9362](https://wikipedia.org/wiki/ISO_9362).

A BIC is a unique identification code for financial and non-financial institutions.
```

The historical background should not be covered in this documentation, instead a link must be added. The target group of the documentation are the users of the components and therefore the use is in the foreground.

codisart replied:
applied

froschdesign commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,42 @@
+
+## BIC validation
+
+BIC codes should be a string which length should be equal to 8 or 11.
```
```suggestion
BICs should be a string which length should be equal to 8 or 11.
```

Otherwise it means "Business Identifier Code codes". 😉

codisart replied:
I have to confess that I have copied the IBAN validator documentation and it uses IBAN number. I will change it

froschdesign commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,42 @@
+
+* The following 2 characters can be letters or digits. It is used to represent a location (like a city)
+
+* The last 3 characters are optional and can be letters or digits. It can represent a branche office. The code 'XXX' is often used to represent the ain office when the 11 characters code is used.
```
Please remove empty lines and reduce the line length. In documentation we follow the coding standard:

> * The soft limit on line length MUST be 120 characters.
> * Lines SHOULD NOT be longer than 80 characters; lines longer than that SHOULD be split into multiple subsequent lines of no more than 80 characters each.

https://www.php-fig.org/psr/psr-12/#23-lines

(https://github.com/laminas/laminas-coding-standard/blob/2.0.0rc1/docs/book/v2/intro.md#laminas-coding-standard)

codisart replied:
applied coding standards

froschdesign commented on docs/book/validators/bic.md:
```diff
@@ -0,0 +1,42 @@
+    // bic appears to be valid
+} else {
+    // bic is invalid; print the reasons
+}
```
```suggestion
if ($validator->isValid('DEUTDEFF')) {
    // Value appears to be valid
}
```

We can use a _real_ example for better demonstration of usage and to allow the example code to be copied.

codisart replied:
added a real example

froschdesign reviewed (changes requested):
Good extension for this component! 👍 

I added some small comments and suggestions.

michalbundyra commented on src/BusinessIdentifierCode.php:
```diff
@@ -0,0 +1,93 @@
+{
+    private const INVALID = 'valueNotBic';
+    private const NOT_STRING = 'valueNotString';
+    private const NOT_VALID_COUNTRY = 'valueNotCountry';
```
As per @froschdesign comment - https://github.com/laminas/laminas-validator/pull/58/files#r400668393 - these constant should be `public` and then we can use them in tests.

codisart replied:
changed

codisart commented:
Hello @froschdesign and @michalbundyra, are my last changes ok with your recommendations ?

codisart commented:
Hello @froschdesign and @michalbundyra, are there changes needed to this pull request ?

codisart commented:
Hello @boesing,

I see that you removed the PR from the 2.14 milestone.
What branch should I use to rebase against now ?
What can I do to help to merge this PR ?

boesing commented:
@codisart I did not dive very deep into this but there are some change requests in here which will probably block the release of 2.14.0.
As #75 will contain PHP 8 support, I'd like to release this component as far as anything regarding PHP 8 is ready.

After the release, we can re-target this branch to 2.15.
However, as stated [here](https://github.com/laminas/laminas-validator/pull/58#discussion_r525606982), we might want to use official documentation as references when applying RegExp.

weierophinney committed:
docs: format for consistency

- 1 line per sentence
- Use `-` for bullets (not `*`)
- Add whitespace after headings

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney committed:
qa: use officially documented regex (with tweaks)

This patch updates the REGEX_BIC constant to follow the documentation from the bundesbank.
The main changes are that the two characters following the country code have more restrictions (the first cannot contain the digit 1, the second cannot contain the letter "o").
A comment with a link to the documentation (and reference to the page) is provided.

CS was also indicating we had two unused import statements, so I removed those.

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>

weierophinney commented:
I updated the REGEX to follow the BundesBank documentation (it was just some minor changes). I also did some cleanup on the documentation.
